### PR TITLE
New component to handle dynamic form creation

### PR DESCRIPTION
**Pull request by pserquiz**
# Critical Changes

# Changes
A component that will handle dynamic form creation based on the given fieldset

# Issues Closed

### CODE REVIEW

File: force-app/main/default/lwc/appFieldSet/appFieldSet.js-meta.xml [hunk 1]
```diff
@@ -10,8 +10,8 @@
 	</targets>
 	<targetConfigs>
 		<targetConfig targets="lightningCommunity__Default">
-			<property name="fieldSetObject" type="String" label="What is the object" description="Which object has the fieldset"/>
-			<property name="fieldSetApiName" type="String" label="What is the fieldSet api" description="The api name of the fieldset that will be used"/>
+			<property name="fieldSetObject" type="String" label="The Object Api Name" description="Which object has the fieldset"/>
+			<property name="fieldSetApiName" type="String" label="The FieldSet Api Name" description="The api name of the fieldset that will be used"/>
 			<property name="isUsedInCommunityBuilder" type="Boolean" label="Check if its being used in the community builder" description="This is component is embed to a parent component?" default="false"/>
```
Comment (cwico) [inline force-app/main/default/lwc/appFieldSet/appFieldSet.js-meta.xml]: Why do you need this? Is this suppose to be the flag for standalone form? If so can you call it isStandalone=true as default.

File: force-app/main/default/classes/AppFieldSetController.cls [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+	}
+
+	/**
+	 * Inserts a generic sobject
+	 *
+	 * @param jSONSObject the object to be inserted
+	 * @param sObjectApiName the object api name
+	 */
+	@AuraEnabled
+	public static void insertSObject(String jSONSObject, String sObjectApiName) {
```
Comment (cwico) [inline force-app/main/default/classes/AppFieldSetController.cls]: some formatting but this looks great.

File: force-app/main/default/lwc/appFieldSet/appFieldSet.js [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+	}
+
+	insertObject() {
+		return insertSObject({jSONSObject: JSON.stringify(this.buildObjectToInsert()), sObjectApiName: this.fieldSetObject}).then(() => {
+			this.handleToastMessage('Success', 'Successfully Inserted', 'success');
+		})
+	}
+
+	//Support methods
+	buildObjectToInsert() {
```
Comment (cwico) [inline force-app/main/default/lwc/appFieldSet/appFieldSet.js]: loooks great! how can we support namespaced field?

Comment (cwico) [reply force-app/main/default/lwc/appFieldSet/appFieldSet.js]: I think we should make this more generic like getRecord()?

File: force-app/main/default/lwc/appFieldSet/appFieldSet.js-meta.xml (hunk 1 above)
Comment (pserquiz) [reply force-app/main/default/lwc/appFieldSet/appFieldSet.js-meta.xml]: thats to be used as design attributes, this component can be dragged and drop in the community builder too.

File: force-app/main/default/lwc/appFieldSet/appFieldSet.js (hunk 3 above)
Comment (pserquiz) [reply force-app/main/default/lwc/appFieldSet/appFieldSet.js]: it builds based on the fields we get from the backend, it should handle namespace already

**Review by cwico — APPROVED** [review]
awesome

**Comment by ghost** [comment]
This is not ready to merge yet.

@pserquiz @cwico to finish up

File: force-app/main/default/classes/AppFieldSetController.cls [hunk 4]
```diff
@@ -0,0 +1,315 @@
...
+	// custom lookup for sfdc-ui-lookup-lwc - https://github.com/pozil/sfdc-ui-lookup-lwc/tree/master/src-sample
+	@AuraEnabled(Cacheable=true)
+	public static List<LookupSearchResult> searchGenericLookup(String searchTerm, List<String> selectedIds, String lookupObject) {
+		// Prepare query parameters
+		searchTerm += '*';
+
+		// Execute search query
+		String query = 'FIND \'' + String.escapeSingleQuotes(searchTerm) + '\' IN ALL FIELDS RETURNING ' + lookupObject + '(Id, Name WHERE Id NOT IN :selectedIds) LIMIT :MAX_RESULTS';
+		system.debug('query => '+query);
+		List<List<SObject>> searchResults = Search.query(query);
```
Comment (cwico) [inline force-app/main/default/classes/AppFieldSetController.cls]: I don't think we need double lists here anymore. Can you take that out?

File: force-app/main/default/classes/AppFieldSetController.cls [hunk 5]
```diff
@@ -0,0 +1,315 @@
...
+		SObject[] filteredResults = (List<SObject>) searchResults[0];
+		for (SObject obj : filteredResults) {
+			String name = (String) obj.get('Name');
+			results.add(
+					new LookupSearchResult(
+							obj.Id,
+							lookupObject,
+							icon,
+							name,
+							lookupObject+' • ' + name
```
Comment (cwico) [inline force-app/main/default/classes/AppFieldSetController.cls]: Remove or comment the **' • ' + name** for now or replace name with id

File: force-app/main/default/classes/AppFieldSetController.cls [hunk 6]
```diff
@@ -0,0 +1,315 @@
...
+	 * @param jSONSObject the object to be converted
+	 * @param sObjectApiName the object api name
+	 */
+	private static SObject convertGenericSObject(String jSONSObject, String sObjectApiName) {
+		Map<String, Object> fieldMap = (Map<String, Object>)JSON.deserializeUntyped(jSONSObject);
+		Schema.SObjectType targetType = Schema.getGlobalDescribe().get(sObjectApiName);
+		SObject newSobject = targetType.newSObject();
+		Map<String, Schema.sObjectField> targetFields = targetType.getDescribe().fields.getMap();
+		for (String key : fieldMap.keySet())
+		{
```
Comment (cwico) [inline force-app/main/default/classes/AppFieldSetController.cls]: braces formatting

File: force-app/main/default/classes/AppFieldSetControllerTest.cls [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+    LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+    ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+    POSSIBILITY OF SUCH DAMAGE.
+*/
+/**
+ * @description Test class for AppFieldSet
+ * @author  Pedro Serquiz, Traction on Demand
+ * @date    2020-04-19
+ */
+@IsTest
```
Comment (cwico) [inline force-app/main/default/classes/AppFieldSetControllerTest.cls]: try Parallel?

File: force-app/main/default/lwc/appFieldSet/appFieldSet.html [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+ -->
+
+<!-- App Field Set -->
+<template>
+	<c-app-spinner></c-app-spinner>
+	<div class="slds-grid slds-wrap slds-gutters">
+		<template for:each={fieldSetData} for:item="field">
+			<div key={field.fieldPath} class="slds-col slds-small-size_1-of-1 slds-medium-size_1-of-2 slds-large-size_1-of-2">
+				<template if:true={field.isLookupField}>
+					<c-lookup
```
Comment (cwico) [inline force-app/main/default/lwc/appFieldSet/appFieldSet.html:12]: yay

File: force-app/main/default/lwc/appFieldSet/appFieldSet.js [hunk 9]
```diff
@@ -0,0 +1,363 @@
...
+
+	@api
+	get record() {
+		return this.buildRecord();
+	}
+
+	renderedCallback() {
+		if (!this.isFirstRender) {
+			this.isFirstRender = true;
+			getOrgNamespace().then(result => {
```
Comment (cwico) [inline force-app/main/default/lwc/appFieldSet/appFieldSet.js:53]: Smart, should we create a base component that we can start using for namespace and other stuff. Namespace for now.

File: force-app/main/default/lwc/appUtils/appUtils.js [hunk 10]
```diff
@@ -24,9 +24,16 @@ const getFieldValue = (obj, fieldName, namespace) => {
...
+		});
+		return newList;
+	} else {
+		return list;
+	}
 }
 
-export {handleToastMessage, getOrgNamespace, getFieldValue, applyNamespace}
\ No newline at end of file
+export {handleToastMessage, getOrgNamespace, getFieldValue, setNamespaceToList}
```
Comment (cwico) [inline force-app/main/default/lwc/appUtils/appUtils.js]: Why replace applyNamespace with setNamespaceToList? when applyNamespace works with lists also and any nested objects and fields. Can you double check this?

File: force-app/main/default/lwc/appFieldSet/appFieldSet.js [hunk 11]
```diff
@@ -0,0 +1,363 @@
...
+
+	handleLookupChange() {
+		this.lookupConfig.lookupErrors = [];
+	}
+
+	validateLookup() {
+		const selection = this.template.querySelector('c-lookup').getSelection();
+		if (selection.length === 0) {
+			this.lookupConfig.lookupErrors = [
+				{ message: 'Please choose your care facility to continue...' }
```
Comment (cwico) [inline force-app/main/default/lwc/appFieldSet/appFieldSet.js]: Can you change this label?  to please choose your "OBJECT with name (if possible)"   to continue

**Review by cwico — CHANGES_REQUESTED** [review]
There are some notes that I put on here. Needs some minor rivisions.

File: force-app/main/default/lwc/appFieldSet/appFieldSet.js (hunk 9 above)
Comment (pserquiz) [reply force-app/main/default/lwc/appFieldSet/appFieldSet.js:53]: Yeah, good idea, but I dont think we need this for now, maybe have a ticket so we dont forget later on.

File: force-app/main/default/lwc/appUtils/appUtils.js (hunk 10 above)
Comment (pserquiz) [reply force-app/main/default/lwc/appUtils/appUtils.js]: sorry, my branch was behind, did a pull and will fix that now, nice catch

Comment (pserquiz) [reply force-app/main/default/lwc/appUtils/appUtils.js]: fixed o/

**Review by cwico — APPROVED** [review]
Looks good

**Pull request merged**